### PR TITLE
FEATURE: SD-1026 Merge modifications made by SD-1012 to SDK8

### PR DESCRIPTION
**Pull request by anvuong119**
---

Supported PR commands:

| Command         | Description            |
| --------------- | ---------------------- |
| `ok to test`    | Re-run standard checks |
| `extended test` | BackstopJS tests       |

---

# PR Checklist

TBD

-   [ ]

# Related Jira tasks

SD-1026: https://jira.intgdc.com/browse/SD-1026

### CODE REVIEW

File: libs/sdk-ui/src/base/measureTitles/ignoreTitlesForSimpleMeasures.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+function isAdhocMeasure(definition: IMeasureDefinitionType): boolean {
+    return (
+        isMeasureDefinition(definition) &&
+        (!!definition.measureDefinition.aggregation ||
+            !!definition.measureDefinition.computeRatio ||
+            (Array.isArray(definition.measureDefinition.filters) &&
+                definition.measureDefinition.filters.length > 0))
+    );
+}
```
Comment (lupko) [inline libs/sdk-ui/src/base/measureTitles/ignoreTitlesForSimpleMeasures.ts:25]: better place for this is the sdk-model package. we should have an exported, specialized type guard for this.

see `isSimpleMeasure` in sdk-model. we can have isAdhocMeasure. export, document, add to barrel. then use here

File: libs/sdk-ui/src/base/measureTitles/ignoreTitlesForSimpleMeasures.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+function isNonSimpleMeasure(definition: IMeasureDefinitionType): boolean {
```
Comment (lupko) [inline libs/sdk-ui/src/base/measureTitles/ignoreTitlesForSimpleMeasures.ts:27]: do we even need this? there is `isSimpleMeasure` type guard. so you can just negate that.

even if we need something like this, again it needs to go into sdk-model.

File: libs/sdk-ui/src/base/measureTitles/ignoreTitlesForSimpleMeasures.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+export function ignoreTitlesForSimpleMeasures<T extends IInsightDefinition>(insight: T): T {
+    const buckets = insightBuckets(insight);
+    // tslint:disable-next-line:no-object-literal-type-assertion
+    return {
+        insight: {
+            ...insight.insight,
+            buckets: buckets.map((bucket: IBucket): IBucket => removeTitleForSimpleMeasure(bucket)),
+        },
+    } as T;
+}
```
Comment (lupko) [inline libs/sdk-ui/src/base/measureTitles/ignoreTitlesForSimpleMeasures.ts:75]: this needs to be generalized and most of the code moved to sdk-model. a model method that aids mass-modification of insight buckets is definitely helpful. instead of this to-the-point approach, we should have something like:

insightModifyItems(insight, modifyFunction).

the method will take care of:
1. iteration over buckets
2. iteration over items
3. calling the modifyFunction

the contract of modifyFunction is, that whatever it returns will be used instead of the original item.

4. create new insight definition with modified buckets

---

ideally, steps 1+2 above would be further handled by bucketsModifyItem(buckets, modifyFunction) and bucketModifyItems(bucket, modifyFunction) so that the modification functionality can be used from multiple places.

---

additionally, please change the title removal code so that it uses the existing modifyMeasure function `modifyMeasure(measure, m => m.noTitle());` << this will drop the title

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/140/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/1724/ : SUCCESS in 11m 34s

File: libs/sdk-ui/src/base/measureTitles/ignoreTitlesForSimpleMeasures.ts (hunk 2 above)
Comment (anvuong119) [reply libs/sdk-ui/src/base/measureTitles/ignoreTitlesForSimpleMeasures.ts]: Thanks, I've updated the PR.

File: libs/sdk-ui/src/base/measureTitles/ignoreTitlesForSimpleMeasures.ts (hunk 1 above)
Comment (anvuong119) [reply libs/sdk-ui/src/base/measureTitles/ignoreTitlesForSimpleMeasures.ts]: Thanks, I've updated the PR.

File: libs/sdk-ui/src/base/measureTitles/ignoreTitlesForSimpleMeasures.ts (hunk 3 above)
Comment (anvuong119) [reply libs/sdk-ui/src/base/measureTitles/ignoreTitlesForSimpleMeasures.ts]: Thanks, I've updated the PR.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/148/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/1735/ : SUCCESS in 11m 50s

File: libs/sdk-model/__mocks__/model.ts [hunk 4]
```diff
@@ -59,6 +59,63 @@ export const WinRate: IMeasure<IMeasureDefinition> = newMeasure("aaX0PIUzg7nF");
...
 export const Won: IMeasure<IMeasureDefinition> = newMeasure("afSEwRwdbMeQ");
+
+/**
+ * Creates an adhoc measure with aggregation "sum"
+ */
+export const AggregateAdhoc = (): IMeasure<IMeasureDefinition> => {
+    const adhocMeasure = newMeasure("afSEwRwdbMeQ");
+    adhocMeasure.measure.definition.measureDefinition.aggregation = "sum";
+    return adhocMeasure;
+};
```
Comment (lupko) [inline libs/sdk-model/__mocks__/model.ts:70]: a more 'native' approach to do this using model functions:

export const AggregateAdhoc = modifySimpleMeasure(Won, (m) => m.aggregation("sum").defaultLocalId())

File: libs/sdk-model/__mocks__/model.ts [hunk 5]
```diff
@@ -59,6 +59,63 @@ export const WinRate: IMeasure<IMeasureDefinition> = newMeasure("aaX0PIUzg7nF");
...
+
+/**
+ * Creates an adhoc measure with empty filters
+ */
+export const EmptyFiltersAdhoc = (): IMeasure<IMeasureDefinition> => {
```
Comment (lupko) [inline libs/sdk-model/__mocks__/model.ts:75]: a more 'native' approach to do this using model functions:

modifySimpleMeasure(Won, (m) => m.noFilters().defaultLocalId())

File: libs/sdk-model/__mocks__/model.ts [hunk 6]
```diff
@@ -59,6 +59,63 @@ export const WinRate: IMeasure<IMeasureDefinition> = newMeasure("aaX0PIUzg7nF");
...
+    const adhocMeasure = newMeasure("afSEwRwdbMeQ");
+    adhocMeasure.measure.definition.measureDefinition.filters = [];
+    return adhocMeasure;
+};
+
+/**
+ * Creates an adhoc measure with non-empty filters
+ */
+export const NonEmptyFiltersAdhoc = (): IMeasure<IMeasureDefinition> => {
```
Comment (lupko) [inline libs/sdk-model/__mocks__/model.ts:84]: a more 'native' approach to do this using model functions:

modifySimpleMeasure(Won, (m) => m.filters(newNegativeAttributeFilter(Account.Name, ['value'])).defaultLocalId())

File: libs/sdk-model/__mocks__/model.ts (hunk 4 above)
Comment (lupko) [reply libs/sdk-model/__mocks__/model.ts:70]: or alternatively, you don't even have to do this and use existing ad-hoc measure such as Duration.Sum

File: libs/sdk-model/__mocks__/model.ts [hunk 7]
```diff
@@ -59,6 +59,63 @@ export const WinRate: IMeasure<IMeasureDefinition> = newMeasure("aaX0PIUzg7nF");
...
+            },
+        },
+    ];
+    return adhocMeasure;
+};
+
+/**
+ * Creates an adhoc measure with a computeRatio of false
+ */
+export const FalseComputeRatioAdhoc = (): IMeasure<IMeasureDefinition> => {
```
Comment (lupko) [inline libs/sdk-model/__mocks__/model.ts:104]: a more 'native' approach to do this using model functions:

modifySimpleMeasure(Won, (m) => m.noRatio().defaultLocalId())

File: libs/sdk-model/__mocks__/model.ts [hunk 8]
```diff
@@ -59,6 +59,63 @@ export const WinRate: IMeasure<IMeasureDefinition> = newMeasure("aaX0PIUzg7nF");
...
+    const adhocMeasure = newMeasure("afSEwRwdbMeQ");
+    adhocMeasure.measure.definition.measureDefinition.computeRatio = false;
+    return adhocMeasure;
+};
+
+/**
+ * Creates an adhoc measure with a computeRatio of true
+ */
+export const TrueComputeRatioAdhoc = (): IMeasure<IMeasureDefinition> => {
```
Comment (lupko) [inline libs/sdk-model/__mocks__/model.ts:113]: a more 'native' approach to do this using model functions:

modifySimpleMeasure(Won, (m) => m.ratio().defaultLocalId())

File: libs/sdk-model/src/execution/buckets/index.ts [hunk 9]
```diff
@@ -423,3 +424,26 @@ export function disableComputeRatio<T extends IAttributeOrMeasure>(item: T): T {
 
     return item;
 }
+
+/**
+ * Describes the type of the function used to modify the bucket items.
+ *
+ * @public
+ */
+export type BucketItemModifier = (bucketItem: IAttributeOrMeasure) => IAttributeOrMeasure;
```
Comment (lupko) [inline libs/sdk-model/src/execution/buckets/index.ts:433]: please rename to `BucketItemModification` (we use similar elsewhere.. MeasureModifications function etc)

File: libs/sdk-model/src/execution/buckets/index.ts [hunk 10]
```diff
@@ -423,3 +424,26 @@ export function disableComputeRatio<T extends IAttributeOrMeasure>(item: T): T {
...
+
+/**
+ * Creates a new bucket with modified bucket items (retrieved by applying the modifyFunction to the bucketItem).
+ *
+ * @param bucket - maybe a simple measure where compute ratio should be disabled
```
Comment (lupko) [inline libs/sdk-model/src/execution/buckets/index.ts:438]: this param description does not look valid :)

File: libs/sdk-model/src/execution/buckets/index.ts [hunk 11]
```diff
@@ -423,3 +424,26 @@ export function disableComputeRatio<T extends IAttributeOrMeasure>(item: T): T {
...
+
+/**
+ * Describes the type of the function used to modify the bucket items.
+ *
+ * @public
+ */
+export type BucketItemModifier = (bucketItem: IAttributeOrMeasure) => IAttributeOrMeasure;
+
+/**
+ * Creates a new bucket with modified bucket items (retrieved by applying the modifyFunction to the bucketItem).
```
Comment (lupko) [inline libs/sdk-model/src/execution/buckets/index.ts:436]: I suggest to reword this as follows:

`Creates a new bucket by modifying items of the provided input bucket. Each item from the input bucket will be dispatched to the modification function and the result of the modification will be included in the new bucket.

Note: it is valid for the modification function to just return the original item. In that case the item will be included in the bucket without modification.
`

File: libs/sdk-model/src/execution/measure/index.ts [hunk 12]
```diff
@@ -203,6 +203,21 @@ export function isSimpleMeasure(obj: any): obj is IMeasure<IMeasureDefinition> {
...
+ * @public
+ */
+export function isAdhocMeasure(obj: any): obj is IMeasure<IMeasureDefinition> {
+    return (
+        isSimpleMeasure(obj) &&
+        (!!obj.measure.definition.measureDefinition.aggregation ||
+            !!obj.measure.definition.measureDefinition.computeRatio ||
+            (Array.isArray(obj.measure.definition.measureDefinition.filters) &&
+                obj.measure.definition.measureDefinition.filters.length > 0))
+    );
```
Comment (lupko) [inline libs/sdk-model/src/execution/measure/index.ts:218]: please simplify / decompose this conditional. i suggest something like this:

`
if (!isSimpleMeasure(obj)) return false;

const {measureDefinition} = obj.measure.definition;

return measureDefinition.aggregation || measureDefinition.computeRatio || ..

`

File: libs/sdk-model/src/execution/measure/index.ts [hunk 13]
```diff
@@ -203,6 +203,21 @@ export function isSimpleMeasure(obj: any): obj is IMeasure<IMeasureDefinition> {
     return isMeasure(obj) && isMeasureDefinition(obj.measure.definition);
 }
 
+/**
+ * Type guard for checking whether object is an adhoc measure.
+ *
```
Comment (lupko) [inline libs/sdk-model/src/execution/measure/index.ts:208]: please expand the documentation so that it describes what is an adhoc measure

File: libs/sdk-model/src/insight/index.ts [hunk 14]
```diff
@@ -579,6 +587,29 @@ export function insightSetBuckets<T extends IInsightDefinition>(
     } as T;
 }
 
+/**
+ * Creates a new insight with modified bucket items (retrieved by applying the modifications function to each bucketItem in the insight).
+ *
```
Comment (lupko) [inline libs/sdk-model/src/insight/index.ts:592]: please expand this doc with something like this:

`Note: the bucket item modification function SHOULD NOT modify bucket item's localId. The localId MAY be used to reference the item from other places in the insight (for example from sorts). Changing the item localId has potential to break the insight: as-is this function does not concern itself with changing the references.`

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/149/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/1737/ : FAILURE in 5m 33s

File: libs/sdk-model/__mocks__/model.ts (hunk 4 above)
Comment (anvuong119) [reply libs/sdk-model/__mocks__/model.ts]: Thanks, I've updated the PR.

File: libs/sdk-model/__mocks__/model.ts (hunk 5 above)
Comment (anvuong119) [reply libs/sdk-model/__mocks__/model.ts]: Thanks, I've updated the PR.

File: libs/sdk-model/__mocks__/model.ts (hunk 6 above)
Comment (anvuong119) [reply libs/sdk-model/__mocks__/model.ts]: Thanks, I've updated the PR.

File: libs/sdk-model/__mocks__/model.ts (hunk 7 above)
Comment (anvuong119) [reply libs/sdk-model/__mocks__/model.ts]: Thanks, I've updated the PR.

File: libs/sdk-model/__mocks__/model.ts (hunk 8 above)
Comment (anvuong119) [reply libs/sdk-model/__mocks__/model.ts]: Thanks, I've updated the PR.

File: libs/sdk-model/src/execution/buckets/index.ts (hunk 9 above)
Comment (anvuong119) [reply libs/sdk-model/src/execution/buckets/index.ts]: Thanks, I've updated the PR.

File: libs/sdk-model/src/execution/buckets/index.ts (hunk 10 above)
Comment (anvuong119) [reply libs/sdk-model/src/execution/buckets/index.ts]: Thanks, I've updated the PR.

File: libs/sdk-model/src/execution/buckets/index.ts (hunk 11 above)
Comment (anvuong119) [reply libs/sdk-model/src/execution/buckets/index.ts]: Thanks, I've updated the PR.

File: libs/sdk-model/src/execution/measure/index.ts (hunk 12 above)
Comment (anvuong119) [reply libs/sdk-model/src/execution/measure/index.ts]: Thanks, I've updated the PR.

File: libs/sdk-model/src/execution/measure/index.ts (hunk 13 above)
Comment (anvuong119) [reply libs/sdk-model/src/execution/measure/index.ts]: Thanks, I've updated the PR.

File: libs/sdk-model/src/insight/index.ts (hunk 14 above)
Comment (anvuong119) [reply libs/sdk-model/src/insight/index.ts:592]: Thanks, I've updated the PR.

**Comment by anvuong119** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/152/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/1741/ : FAILURE in 4m 23s

**Comment by anvuong119** [comment]
ok to test

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/155/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/1744/ : FAILURE in 4m 29s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/157/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/1746/ : SUCCESS in 12m 02s

**Comment by lupko** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/194/ : SUCCESS in 21s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/449/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/1770/ : SUCCESS in 11m 48s
